harness/tests/cli/test_auth.py

```diff
@@ -43,9 +43,263 @@ def test_auth_with_store(requests_mock: requests_mock.Mocker, user: Optional[str
+        return (
+            f"ScenarioSet({self.req_user!r}, {self.req_pass!r}, {self.env_user!r}, "
+            f"{self.env_pass!r}, {self.env_token!r}, {self.cache!r}, {results})"
+        )
+
+    def scenarios(self) -> Iterator[Tuple[str, str, str, str, str, str]]:
+        for req_user in "yn" if self.req_user == "*" else self.req_user:
+            for req_pass in "yn" if self.req_pass == "*" else self.req_pass:
+                for env_user in "yne" if self.env_user == "*" else self.env_user:
+                    for env_pass in "yn" if self.env_pass == "*" else self.env_pass:
+                        for env_token in "yn" if self.env_token == "*" else self.env_token:
+                            for cache in "ynx" if self.cache == "*" else self.cache:
+                                yield req_user, req_pass, env_user, env_pass, env_token, cache
+
+
+@pytest.mark.parametrize(
+    "scenario_set",
+    [
+        # Explicit user and password, no environment settings, or DET_USER is overridden by
+        # explicitly requested user.
+        #
+        #           requested user (y="user", n=None)
+        #            |   requested password (y="req_pass", n=None)
+        #            |    |   DET_USER env var (y="user", n=unset, e="extra")
+        #            |    |    |    DET_PASS env var (y="env_pass, e=None)
+        #            |    |    |     |   DET_USER_TOKEN (y="env_token", e=None)
+        #            |    |    |     |    |   token cache state (y="cache", n=empty, x="expired")
+        #            |    |    |     |    |    |   Results...
+        #            |    |    |     |    |    |    |
+        ScenarioSet("y", "y", "ne", "*", "*", "y", Check("cache")),
+        ScenarioSet("y", "y", "ne", "*", "*", "n", Login("user", "req_pass")),
+        ScenarioSet("y", "y", "ne", "*", "*", "x", Check("expired"), Login("user", "req_pass")),
+        # ---
+        # Explicit user, but password not provided.  Still no (relevant) environment settings.
+        ScenarioSet("y", "n", "ne", "*", "*", "y", Check("cache")),
+        ScenarioSet("y", "n", "ne", "*", "*", "n", Login("user", "prompt_pass")),
+        ScenarioSet("y", "n", "ne", "*", "*", "x", Check("expired"), Login("user", "prompt_pass")),
+        # ---
+        # DET_USER_TOKEN is overridden by a configured cache (the on-cluster `det user login` case).
+        ScenarioSet("*", "*", "ye", "*", "y", "y", Check("cache")),
+        # DET_USER_TOKEN can be used if user is explicit, so long as it matches.
+        ScenarioSet("y", "*", "y", "*", "y", "n", Use("env_token")),
+        # Token in env is used if cache is missing or invalid.
+        ScenarioSet("n", "*", "y", "*", "y", "n", Use("env_token")),
+        ScenarioSet("n", "*", "y", "*", "y", "x", Check("expired"), Use("env_token")),
+        # Token in env is overridden by a configured cache (the on-cluster `det user login` case).
+        ScenarioSet("n", "*", "ye", "*", "y", "y", Check("cache")),
```

Comment:
For cases like this where we use `Check("cache")`, it could also be useful to add `Use("cache")` just to ensure the final result of the returned token.
This could be useful just for the sake of ensuring that the token for the cached user gets returned at the end of the authentication logic.